File: android/src/main/kotlin/com/hellacloud/laikangning_sdk/PluginController.kt
package com.hellacloud.laikangning_sdk

import android.content.Context
import android.util.Log
import com.hellacloud.laikangning_sdk.handlers.LmtpFhrDataChangedHandler
import com.hellacloud.laikangning_sdk.lmtp.ReactiveLmtpClient
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import com.hellacloud.laikangning_sdk.model.ProtobufModel as pb


class PluginController {

    private val pluginMethods =
        mapOf<String, (call: MethodCall, result: MethodChannel.Result) -> Unit>(
            "initialize" to this::initializeClient,
            "deinitialize" to this::deinitializeClient,
            "stopWork" to this::stopWork,
            "startWork" to this::startWork,
            "beginRecordWave" to this::beginRecordWave,
            "finishRecordWave" to this::finishRecordWave,
            "putData" to this::putData,
            "sendTocoReset" to this::sendTocoReset,
            "setFM" to this::setFM,
            "putFhrVolume" to this::putFhrVolume,
        )

    companion object {
        private const val READ_FHR_CHANGE_CHANNEL = "flutter_laikangyun_read_fhr_change_channel"
    }

    private lateinit var mContent: Context
    private var mClient: ReactiveLmtpClient? = null
    private var mFhrDataChangeHandler: LmtpFhrDataChangedHandler? = null

    internal fun initialize(messenger: BinaryMessenger, context: Context) {
        if (mClient != null) {
            Log.i(PluginController::class.java.name, "reinitialize")
            return
        }
        mClient = ReactiveLmtpClient()
        mContent = context

        mFhrDataChangeHandler = LmtpFhrDataChangedHandler()
        val fhrChangeChannel = EventChannel(messenger, READ_FHR_CHANGE_CHANNEL)
        fhrChangeChannel.setStreamHandler(mFhrDataChangeHandler)
    }

    internal fun deinitialize() {
        mClient?.deinitialClient()
        mClient = null
    }

    internal fun execute(call: MethodCall, result: MethodChannel.Result) {
        pluginMethods[call.method]?.invoke(call, result) ?: result.notImplemented()
    }

    // 初始化客户端
    private fun initializeClient(call: MethodCall, result: MethodChannel.Result) {
        mClient?.initializeClient(mFhrDataChangeHandler)
        result.success(null)
    }

    // 释放资源
    private fun deinitializeClient(call: MethodCall, result: MethodChannel.Result) {
        this.deinitialize()
        result.success(null)
    }

    // 停止工作
    private fun stopWork(call: MethodCall, result: MethodChannel.Result) {
        mClient?.stopWork()
        result.success(null)
    }

    // 开始工作
    private fun startWork(call: MethodCall, result: MethodChannel.Result) {
        mClient?.startWork()
        result.success(null)
    }

    // 开启胎心声音记录
    private fun beginRecordWave(call: MethodCall, result: MethodChannel.Result) {
        val recordWave = pb.LknRecordWave.parseFrom(call.arguments as ByteArray)
        mClient?.beginRecordWave(recordWave.path, recordWave.fileName)
        result.success(null)
    }

    // 完成胎心声音记录
    private fun finishRecordWave(call: MethodCall, result: MethodChannel.Result) {
        mClient?.finishRecordWaves()
        result.success(null)
    }

    // 写入蓝牙数据
    private fun putData(call: MethodCall, result: MethodChannel.Result) {
        val lknPutData = pb.LknPutData.parseFrom(call.arguments as ByteArray)
        mClient?.putData(lknPutData.data.toByteArray())
        result.success(null)
    }

    // 设置宫缩复位
    private fun sendTocoReset(call: MethodCall, result: MethodChannel.Result) {
        val lknPutData = pb.LknTocoReset.parseFrom(call.arguments as ByteArray)
        mClient?.sendTocoReset(lknPutData.value)
        result.success(null)
    }

    // 设置一次手动胎动
    private fun setFM(call: MethodCall, result: MethodChannel.Result) {
        mClient?.sendFM()
        result.success(null)
    }

    // 设置胎心音量
    private fun putFhrVolume(call: MethodCall, result: MethodChannel.Result) {
        val lknPutData = pb.LknFhrVolume.parseFrom(call.arguments as ByteArray)
        mClient?.sendFhrVolume(lknPutData.value)
        result.success(null)
    }

}

File: ios/Classes/plugin/PluginController.swift
//
// Created by 候帅 on 2022/10/18.
//

import Foundation
import CoreBluetooth
import Flutter

final class PluginController {

    private var decoder: LMTPDecoder?
    private var streamDelegate: StreamDelegate?

    func initialize(completion: @escaping PlatformMethodCompletionHandler) {
        // 进行相应的初始化逻辑
        guard decoder != nil else {
            decoder = LMTPDecoder.shareInstance() as? LMTPDecoder
            // 添加对应的声明
            return
        }
        completion(.success(nil))
    }

    func deinitialize(completion: @escaping PlatformMethodCompletionHandler) {
        guard let decoder = decoder else {
            print("not initialize")
            return
        }
        decoder.stopMoniter()
        completion(.success(nil))
    }

    func startWork(completion: @escaping PlatformMethodCompletionHandler) {
        decoder?.startRealTimeAudioPlyer()
        completion(.success(nil))
    }

    func setStreamDelegate(_ handler: StreamDelegate) {
        streamDelegate = handler
    }

    func stopWork(completion: @escaping PlatformMethodCompletionHandler) {
        decoder?.stopRealTimeAudioPlyer()
        completion(.success(nil))
    }

    func putData(args data: LknPutData, completion: @escaping PlatformMethodCompletionHandler) {
        let data = decoder?.start(withCharacterData: data.data)
        // 接口传递当前至内容 并推送至 流数据中。
        guard let lknHeaderData = data else {
            completion(.success(nil))
            return
        }
        streamDelegate?.fhrData(withData: lknHeaderData)
        completion(.success(nil))
    }

    func beginRecordWave(args data: LknRecordWave, completion: @escaping PlatformMethodCompletionHandler) {
        let fileManager = FileManager()
        if !fileManager.fileExists(atPath: data.path) {
            do {
                try fileManager.createDirectory(atPath: data.path, withIntermediateDirectories: true)
            } catch {
                print("file manager is error")
            }
        }
        decoder?.startMonitor(withAudioFilePath: "\(data.path)/\(data.fileName)")
        completion(.success(nil))
    }

    func finishRecordWaves(completion: @escaping PlatformMethodCompletionHandler) {
        decoder?.stopMoniter()
        completion(.success(nil))
    }

    func sendTocoReset(args data: LknTocoReset, completion: @escaping PlatformMethodCompletionHandler) {

        completion(.success(nil))
    }

    func sendFM(completion: @escaping PlatformMethodCompletionHandler) {
        decoder?.setFM()
        completion(.success(nil))
    }

    func sendFhrVolume(args data: LknFhrVolume, completion: @escaping PlatformMethodCompletionHandler) {
        completion(.success(nil))
    }
}
